test/raft/unit/test_snapshot.c

```diff
@@ -407,79 +524,242 @@ static void pool_rpc_to_expired(struct rpc *rpc)
 	progress();
 }
 
-/* TODO(alberto): combine them with tests above once the rest is in place.
- * Dispatch to one of two implementations ut or pool in general functions. */
-TEST(snapshot_leader, pool_timeouts, set_up, tear_down, 0, NULL) {
+static bool pool_is_main_thread_op(void) {
+	return thread_identifier == MAGIC_MAIN_THREAD;
+}
+
+static void pool_ht_create_op(pool_work_t *w)
+{
+	if (global_fixture.is_leader) {
+		PRE(!global_fixture.leader.ops->is_main_thread());
+	} else {
+		PRE(!global_fixture.follower.ops->is_main_thread());
+	}
+	(void)w;
+}
+
+static void pool_fill_ht_op(pool_work_t *w)
+{
+	if (global_fixture.is_leader) {
+		PRE(!global_fixture.leader.ops->is_main_thread());
+	} else {
+		PRE(!global_fixture.follower.ops->is_main_thread());
+	}
+	(void)w;
+}
+
+static void pool_write_chunk_op(pool_work_t *w)
+{
+	struct work *work = CONTAINER_OF(w, struct work, pool_work);
+	struct follower *follower = CONTAINER_OF(work, struct follower, work);
+	PRE(!follower->ops->is_main_thread());
+}
+
+static void pool_read_sig_op(pool_work_t *w)
+{
+	struct work *work = CONTAINER_OF(w, struct work, pool_work);
+	struct follower *follower = CONTAINER_OF(work, struct follower, work);
+	PRE(!follower->ops->is_main_thread());
+}
+
+struct uv_sender_send_data {
+	struct sender *s;
+	sender_cb_op cb;
+};
+
+void uv_sender_send_cb(uv_work_t *req) {
+	(void)req;
+}
+
+void uv_sender_send_after_cb(uv_work_t *req, int status) {
+	(void)req;
+	(void)status;
+	global_fixture.msg_valid = true;
+	struct uv_sender_send_data *data = req->data;
+	data->cb(data->s, 0);
+}
+
+int uv_sender_send_op(struct sender *s,
+		struct raft_message *payload,
+		sender_cb_op cb) {
+	/* We only expect one message to be in-flight. */
+	static uv_work_t req;
+	static struct uv_sender_send_data req_data;
+
+	global_fixture.last_msg_sent = *payload;
+	/* Flag is only toggled when the after_cb is called, emulating the message
+	 * being sent. */
+	global_fixture.msg_valid = false;
+	s->cb = cb;
+	req_data = (struct uv_sender_send_data) {
+		.s = s,
+		.cb = cb,
+	};
+	req = (uv_work_t) {
+		.data = &req_data,
+	};
+	uv_queue_work(uv_default_loop(), &req, uv_sender_send_cb, uv_sender_send_after_cb);
+	return 0;
+}
+
+struct raft_message uv_get_msg_sent(void) {
+	munit_assert(global_fixture.msg_valid);
+	global_fixture.msg_valid = false;
+	return global_fixture.last_msg_sent;
+}
+
+TEST(snapshot_leader, pool_timeouts, pool_set_up, pool_tear_down, 0, NULL) {
 	struct leader_ops ops = {
 		.to_init = pool_to_init_op,
 		.to_stop = pool_to_stop_op,
 		.to_start = pool_to_start_op,
-		.ht_create = ut_ht_create_op,
-		.work_queue = ut_work_queue_op,
-		.sender_send = ut_sender_send_op,
+		.ht_create = pool_ht_create_op,
+		.work_queue = pool_work_queue_op,
+		.sender_send = uv_sender_send_op,
+		.is_main_thread = pool_is_main_thread_op,
```

Comment:
It's better not to override `is_main_thread()`.
You have an implementation of `pool_is_main_thread_op()`
Maybe, it's worth thinking about an implementation designed around `uv_key_create()`. We need to check if it's a good idea to mix `__tls` and `uv_`